arduino/libraries/libraries_test.go

```diff
@@ -85,3 +88,89 @@ func TestLibrariesLoader(t *testing.T) {
 		require.True(t, lib.IsLegacy)
 	}
 }
+
+func TestSymlinkLoop(t *testing.T) {
+	// Set up directory structure of test library.
+	libraryPath := paths.New("testdata", "TestLib")
+	tempLibPath, err := paths.TempDir().MkTempDir("TestSymlinkLoop")
+	require.NoError(t, err)
+	defer tempLibPath.RemoveAll() // Clean up after the test.
+	testLib := tempLibPath.Join("TestLib")
+	require.NoError(t, libraryPath.CopyDirTo(testLib))
+	examplesPath := testLib.Join("examples")
+	require.NoError(t, examplesPath.Mkdir())
+
+	// It's probably most friendly for contributors using Windows to create the symlinks needed for the test on demand.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer1").String())
+	require.NoError(t, err, "This test must be run as administrator on Windows to have symlink creation privilege.")
+	// It's necessary to have multiple symlinks to a parent directory to create the loop.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer2").String())
+	require.NoError(t, err)
+
+	// The failure condition is Load() never returning, testing for which requires setting up a timeout.
+	done := make(chan bool)
+	go func() {
+		_, err = Load(testLib, User)
+		done <- true
+	}()
+
+	noHang := assert.Eventually(
+		t,
+		func() bool {
+			select {
+			case <-done:
+				return true
+			default:
+				return false
+			}
+		},
+		20*time.Second,
+		10*time.Millisecond,
+		"Infinite symlink loop while loading library",
+	)
+	require.True(t, noHang)
+	require.Error(t, err)
+}
+
+func TestLegacySymlinkLoop(t *testing.T) {
+	// Set up directory structure of test library.
+	libraryPath, err := paths.TempDir().MkTempDir("TestSymlinkLoop")
+	defer libraryPath.RemoveAll() // Clean up after the test.
+	require.NoError(t, err)
+	err = libraryPath.Join("TestSymlinkLoop.h").WriteFile([]byte{})
+	require.NoError(t, err)
+	examplesPath := libraryPath.Join("examples")
+	err = examplesPath.Mkdir()
+	require.NoError(t, err)
+
+	// It's probably most friendly for contributors using Windows to create the symlinks needed for the test on demand.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer1").String())
+	require.NoError(t, err, "This test must be run as administrator on Windows to have symlink creation privilege.")
+	// It's necessary to have multiple symlinks to a parent directory to create the loop.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer2").String())
+	require.NoError(t, err)
```

Comment:
Can't we save the test library directly in `testdata` directly in git instead of re-creating it?

Reply:
The reason for generating the symlinks on the fly is that, by default, symlinks are only supported on Windows when in admin mode. So if a Windows user clones a repo that contains symlinks when not in admin mode, the symlinks are not generated (you only get a file that contains the target path as text instead of the link).

For this reason, I think it is more contributor friendly to generate symlinks on the fly. The contributor still needs to run the unit tests as administrator, but the test will clearly communicate the requirement to them if they try to run it without doing so.

But that is only about the symlinks. The symlinks could be added on the fly to a library from testdata instead of generating the whole thing.